packages/squiggle-lang/src/PointSet/Discrete.ts

```diff
@@ -15,55 +15,83 @@ import { MixedShape } from "./Mixed";
 
 export class DiscreteShape implements PointSet<DiscreteShape> {
   readonly xyShape: XYShape.XYShape;
-  readonly integralSumCache?: number;
-  readonly integralCache?: Continuous.ContinuousShape;
+
+  // readonly for external functions through accessor fields
+  private _integralSumCache?: number;
+  private _integralCache?: Continuous.ContinuousShape;
 
   constructor(args: {
     xyShape: XYShape.XYShape;
     integralSumCache?: number;
     integralCache?: ContinuousShape;
   }) {
     this.xyShape = args.xyShape;
-    this.integralSumCache = args.integralSumCache;
-    this.integralCache = args.integralCache;
+    this._integralSumCache = args.integralSumCache;
+    this._integralCache = args.integralCache;
+  }
+
+  get integralCache() {
+    return this._integralCache;
+  }
+  get integralSumCache() {
+    return this._integralSumCache;
+  }
+
+  withIntegralSum(integralSumCache: number): DiscreteShape {
+    return new DiscreteShape({
+      xyShape: this.xyShape,
+      integralSumCache,
+      integralCache: this.integralCache,
+    });
   }
 
   shapeMap(fn: (shape: XYShape.XYShape) => XYShape.XYShape): DiscreteShape {
     return new DiscreteShape({
       xyShape: fn(this.xyShape),
+      // FIXME - this seems wrong
```

Comment:
Do you want to fix before the pr, or after?

Reply:
Later.

This part existed before this PR, and I made sure that `shapeMap` is currently called only in `shapeMap().mean()` chain where it doesn't matter.

It creates some risks of future bugs, but I hope to refactor this part before we trigger this bug by accident.